microsetta_admin/server.py

```diff
@@ -512,35 +514,100 @@ def new_kits():
                                **build_login_variables())
 
     elif request.method == 'POST':
-        num_kits = int(request.form['num_kits'])
-        num_samples = int(request.form['num_samples'])
         prefix = request.form['prefix']
         selected_project_ids = request.form.getlist('project_ids')
-        payload = {'number_of_kits': num_kits,
-                   'number_of_samples': num_samples,
-                   'project_ids': selected_project_ids}
+        num_kits = int(request.form['number_of_kits'])
+        num_samples = int(request.form['number_of_samples'])
+
+        generate_barcodes = []
+        i = 1
+        while i <= num_samples:
+            barcode = request.form.get(f'generate_barcodes_sample_{i}')
+            if barcode:
+                generate_barcodes.append([barcode] * num_kits)
+            else:
+                generate_barcodes.append([])
+            i += 1
+
+        user_barcodes = []
+        total_provided_samples = 0
+
+        for i in range(0, 25):
+            barcode_file = request.files.get(f'upload_csv_{i}')
+            if barcode_file:
+                barcode_file = barcode_file.read().decode('utf-8')
+                new_barcodes = [line.split(',')[0].strip('\ufeff\r')
+                                for line in barcode_file.split('\n')
+                                if line.strip()]
+                if len(new_barcodes) != num_kits:
+                    if len(new_barcodes) > num_kits:
+                        error_message = (f'Too many barcodes provided in '
+                                         f'file {barcode_file}.')
+                    else:
+                        error_message = (f'Not enough barcodes provided in '
+                                         f'file {barcode_file}.')
+                    return render_template('create_kits.html',
+                                           error_message=error_message,
+                                           projects=projects,
+                                           **build_login_variables())
+
+                user_barcodes.append(new_barcodes)
+                total_provided_samples += len(new_barcodes)
+
+        # Calculate remaining samples to be generated
+        total_needed_samples = num_kits * num_samples
+        remaining_samples_to_generate = total_needed_samples - \
+            total_provided_samples
+
+        if remaining_samples_to_generate < 0:
+            error_message = (f'Too many samples provided '
+                             f'({total_provided_samples}) '
+                             f'for the specified number of kits and samples '
+                             f'({total_needed_samples}).')
+            return render_template('create_kits.html',
+                                   error_message=error_message,
+                                   projects=projects,
+                                   **build_login_variables())
```

Comment:
Given the fact that the code should be enforcing the number of provided samples on a per-slot basis, does this block still serve a purpose?